tests/resource/test_resource.py

```diff
@@ -0,0 +1,149 @@
+        with pytest.raises(TypeError, match="Incorrect type of input,"):
+            Resources(*params)
+
+    value_error = ((0, -1, {"Hadamard": 1}, 2, -3),)
+
+    @pytest.mark.parametrize("params", value_error)
+    def test_init_value_error(self, params):
+        """Test that a value error is raised if an attribute is initialized with the wrong type."""
+        with pytest.raises(ValueError, match="Incorrect value of input,"):
+            Resources(*params)
+
+    def test_set_attributes_error(self):
+        """Test that an error is raised if we try to set any attribute."""
+        r = Resources()
+        attr_lst = ["num_wires", "num_gates", "depth", "shots", "gate_types"]
+        message = (
+            "<class 'pennylane.resource.resource.Resources'> object does not support assignment"
+        )
+
+        for attr_name in attr_lst:
+            with pytest.raises(AttributeError, match=message):
+                setattr(r, attr_name, 1)
+
+    test_str_data = (
+        ("wires: 0\n" + "gates: 0\n" + "depth: 0\n" + "shots: 0\n" + "gate_types: \n" + "{}"),
+        ("wires: 5\n" + "gates: 0\n" + "depth: 0\n" + "shots: 0\n" + "gate_types: \n" + "{}"),
+        (
+            "wires: 1\n"
+            + "gates: 3\n"
+            + "depth: 3\n"
+            + "shots: 10\n"
+            + "gate_types: \n"
+            + "{'Hadamard': 1, 'PauliZ': 2}"
+        ),
+        (
+            "wires: 4\n"
+            + "gates: 2\n"
+            + "depth: 2\n"
+            + "shots: 100\n"
+            + "gate_types: \n"
+            + "{'Hadamard': 1, 'CNOT': 1}"
+        ),
+    )
+
+    @pytest.mark.parametrize("r, rep", zip(resource_quantities, test_str_data))
+    def test_to_str(self, r, rep):
+        """Test the string representation of a Resources instance."""
+        assert str(r) == rep
+
+    test_rep_data = (
+        "<Resource: wires=0, gates=0, depth=0, shots=0, gate_types={}>",
+        "<Resource: wires=5, gates=0, depth=0, shots=0, gate_types={}>",
+        "<Resource: wires=1, gates=3, depth=3, shots=10, "
+        "gate_types={'Hadamard': 1, 'PauliZ': 2}>",
+        "<Resource: wires=4, gates=2, depth=2, shots=100, "
+        "gate_types={'Hadamard': 1, 'CNOT': 1}>",
+    )
+
+    @pytest.mark.parametrize("r, rep", zip(resource_quantities, test_rep_data))
+    def test_repr(self, r, rep):
+        """Test the repr method of a Resources instance looks as expected."""
+        assert repr(r) == rep
+
+    def test_eq(self):
+        """Test that the equality dunder method is correct for Resources."""
+        r1 = Resources(4, 2, {"Hadamard": 1, "CNOT": 1}, 2, 100)
+        r2 = Resources(4, 2, {"Hadamard": 1, "CNOT": 1}, 2, 100)
+        r3 = Resources(4, 2, {"CNOT": 1, "Hadamard": 1}, 2, 100)  # all equal
+
+        r4 = Resources(1, 2, {"Hadamard": 1, "CNOT": 1}, 2, 100)  # diff wires
+        r5 = Resources(4, 1, {"Hadamard": 1, "CNOT": 1}, 2, 100)  # diff num_gates
```

Comment:
I feel like this case where the number of gates doesn't match the summed values in the `gate_types` dict should get overwritten in the init method or raise an error.